### ISSUE TITLE
Extension stops working and loads CPU

### ISSUE THREAD

**Issue by helperman** [Denis-net/vscode-ahk-manager#16]
I have a high CPU load while working with ahk-scripts using VS Code and your extension.
![379](https://user-images.githubusercontent.com/8351863/67433072-239da180-f5f0-11e9-8101-7ffab340a9de.png)

What's even worse, the most crucial extension's commands stop working.
For example, I launch command "AHK: Run script" and nothing happens. Script doesn't run.
If I close and open VS Code again, the same script runs successfully.
But a few minutes later the issue occurs again. 

I can reproduce the issue working for some time in VS Code with my scripts.
I created and uploaded a CPU profile as mentioned there: https://github.com/Microsoft/vscode/wiki/Performance-Issues

[CPU-20191023T202355.188Z.cpuprofile.zip](https://github.com/Denis-net/vscode-ahk-manager/files/3764433/CPU-20191023T202355.188Z.cpuprofile.zip)

Is it helpful for you to troubleshoot?
Why does it happen and how to fix it?

**Comment by Denis-net** [comment]
Hey @helperman 😊👋🏻

I'm sorry for your situation.
I want to help you.

First of all:
What other extension do you have installed ?

I have discovered that [vinnyjames' AutoHotkey++](https://github.com/vinnyjames/vscode-autohotkey) has issue with its SignatureHelper, that is the only code that can run to check whether you have used a command elsewhere.

Don't know if its your case.

Try to open the vs-code developer tools to see if any message appears.
Try to reproduce the first gif text, and if you can't complete it, is definitely for the above extension.
Now i'll look the log.

**Comment by Denis-net** [comment]
![immagine](https://user-images.githubusercontent.com/45769702/67516980-e265ca00-f6a1-11e9-8669-dd0aa319824a.png)

You have definitely vinnyjames's extension 😜
I have signaled it to the developer 4 days ago in the only way i found, 
if you want to join me:
https://marketplace.visualstudio.com/items?itemName=vinnyjames.vscode-autohotkey-vj&ssr=false#review-details

Try [this extension](https://marketplace.visualstudio.com/items?itemName=rafaelcavasani.vscode-autohotkey-plus-slimcode) instead for syntax color and the formatter.

Anyway, if your issue is that VSC's editor becames readonly is because of SignatureHelper that goes into a loop while searching a "," after the second "MsgBox," for example. 
Try to disable it, reload VSC and then you'll see that VSC will be faster as usual.

Another extension that will rise the CPU on HTML files is  [Auto Rename Tag](https://marketplace.visualstudio.com/items?itemName=formulahendry.auto-rename-tag).

The ideas are awesome, but the implementations less.

If you still have the issue, please tell me more.

**Comment by helperman** [comment]
Hi Denis!
I installed all extensions I've found by "AutoHotkey" keyword and haven't yet got deep inside.
I joined your bug report about the extension.
You're so kind! I highly appreciate your help!
I wish everybody on the Earth would be so cooperative! And we would live in a much better world definitely.
Looks like you're an experienced developer!
I would like to enrich my knowledge and looking for something new to learn.
Do you have our own web page, youtube channel or another resource for learning about you, your experience in AHK or programming using VS Code? Maybe best practice, tips and tricks, and another essential material to learn.

**Comment by Denis-net** [comment]
Hey Anthony !

Thank you too !
I'm happy that it didn't happened only to me 😂

I have took some time to reply, because i needed to find the right words.
Don't take them too seriously(🤡), are just written for fun 👍🏻

First of all: 
I like your enthusiasm !
i think is a very good thing, but you must remember that you can't rely only on that because emotions even if they can drive you, they can't last, as soon as you encounter an issue (a difficult one) you may lose it and then give up.
Emotions can inspire you, but as they come, they can goes away.
What I'm trying to tell you is that I'm enthusiastic too, but i use "consistency" and a "book that is titled `Don't Panic` in the front with reassuring fonts"(🤣) in this kind of stuff.

I'm nothing special, just a consistent player, who have found time to code for the shake to code and learn 😉.

Yes, I do what i do because i like it, i feel like i'm playing all the time.
So another important thing you must understand in order to proceed is:
Is this (programming) what you really want to do for the rest of your days ?
Because you need motivation and if you feel like you are wasting time you probably are.
If you can clarify it, 
you then know that you do it for a reason: you have a goal or you need to find it in order to proceed.
If you feel yourself on a mission, you know you can't step back and give up, because (i used to think, mistakenly but it works) "If i don't do it, nobody  will do  and that idea that i had simply will remain just that an idea (maybe one day, but that day i could simply don't be there)"
And was more or less the background idea that drive me with  this extension.

When I started to develop VSC's extensions, at that time, there was just one great extension: the  [stef-levesque one](https://marketplace.visualstudio.com/items?itemName=slevesque.vscode-autohotkey)

I started my programmer journey at ITIS school, but an high school has slow teaching times ( depending by some teachers and some students), so I started looking things from myself.
Yes: everybody needs solid bases, but can't pretend to reach something if he just follow others, because he will be simply where the teacher can go (he have to follow a school plan).
And if you are curious ?
You can't simply go where others want and complain if is not where you want to go.

Then other AHK extension comes one by one.

I think this happened because they(we) have seen that is possible.

When I begin with VSC i was like you: i installed all extensions that i found intresting in batch  but when i came across the html issue i had to stop and disable one by one to check what it was wrong with my editor.
And that was frustrating, but from this experience i learned an important lesson: **go light and always know what you have in your pockets**.

TL;DR; Always have a goal in your mind, plan, study the 'unknown known unknown', and make it happens.

For the planning i highly reccommend [Microsoft Todo](https://todo.microsoft.com/) or Keep a noteblock or whatever can help you write down a map made up by steps, so you always know what to do now/next.

When learning is important to focus yourself on a running project, because i found myself lose time studing things that i couldn't apply so i forgot them.

Unfortunately no: i don't have this kind of stuff.
I keep myself **simple** and what i'm always tring to do is automations.
If i think a particular aspect can be made simpler, with fewer errors or more faster, i start to think what program can solve it, and if i found it i use it, otherwise i roll up my sleves and i start throwing some code, and if it works i proceed with refinements.
I think it's just that.

When approacing new languages the first thing to do is open the manual, docs, getting-started and then playing with it, to see what it has to offer.

I discovered that AHK does a great work in hotkeys detections (even in Desktop) that for example in C or C# you will require more and more effort for intercepting a simple hotkey outside your program (called hooks).
When you start crafting more scripts you start to see chaos in your tray icon, because each script can have its tray menu controller.
This is when my extension come in rescue and allow you interact with running script, so you don't have to deal with the chaos.

If you want more material i have some intresting sites to check:
For VS Code:
https://vscodecandothat.com/
https://vscode.rocks/
(and these if you want to craft VSC extensions)
https://code.visualstudio.com/docs
https://code.visualstudio.com/api

AHK:
https://www.autohotkey.com/docs/Program.htm

The Essential material:
https://www.cin.ufpe.br/~cavmj/104The%20Pragmatic%20Programmer,%20From%20Journeyman%20To%20Master%20-%20Andrew%20Hunt,%20David%20Thomas%20-%20Addison%20Wesley%20-%201999.pdf
This is the best book i have ever read,
and the most amazing thing is that it's from 1999 and it is still current in topics. 

Anyway today i started looking the vinnyjames ahk extension source code and i found [this intresting forum](https://www.autohotkey.com/boards/viewtopic.php?t=6033) about OOP in AHK.
Honestly i don't use AHK for big projects, because my primary language is C# which allows me to build fast and reliables programs.
But in the forum they explain very well the argument.

What about you ?
What do you like to code ?
Have you interesting ideas that are waiting for you to craft them ?

Sorry, I wrote a bit 😅

**Comment by Denis-net** [comment]
I didn't mean to offend you, 
I'm terribly sorry.😔

Anyway,
I have fixed the issue and then opened a [PR](https://github.com/cweijan/vscode-autohotkey/pull/3) in the vinnyjames' repo.
Hope he will find it helpful and will find time to merge the fix or generating new ideas. 🤞🏻

This is the best i can do.
May you have a nice day !

**Comment by Denis-net** [comment]
He did it ! 🎊🎉

Happy Halloween !
